code/game/machinery/turrets/turrets.dm

```diff
@@ -0,0 +1,225 @@
+// Base type for rewritten turrets, designed to hopefully be cleaner and more fun to use and fight against.
+// Use the subtypes for 'real' turrets.
+
+/obj/machinery/turret
+	name = "don't use me"
+	icon = 'icons/obj/turrets.dmi'
+	icon_state = "turret_barrel"
+	anchored = TRUE
+	idle_power_usage = 500
+	active_power_usage = 60000 // This determines how fast the internal cell gets charged.
+
+	transform_animate_time = 0.2 SECONDS
+
+	// Visuals.
+	var/image/turret_stand = null
+	var/image/turret_ray = null
+	var/ray_color = "#FFFFFFFF" // Color of the ray, changed by the FSM when switching states.
+
+	// Sounds
+	var/turn_on_sound = null // Played when turret goes from off to on.
+	var/turn_off_sound = null // The above, in reverse.
+
+	// Shooting
+	var/obj/item/gun/installed_gun = null // Instance of the gun inside the turret.
+	var/gun_type = /obj/item/gun/energy/laser/practice // The type of gun to put in the above var on init.
+	var/gun_looting_prob = 25 // If the turret dies and then is disassembled, this is the odds of getting the gun.
+
+	// Power
+	var/obj/item/cell/internal_cell = null // The battery that is used to fire the gun and for the turret to work semi-independantly of the powernet.
+	var/cell_type = /obj/item/cell/high // The type this turret will have on initialization.
+	var/can_charge_cell = TRUE // If FALSE, the battery has to be swapped out manually.
+	var/can_swap_cell = FALSE // If TRUE, the cell can be replaced with a screwdriver.
+	var/cell_looting_prob = 25 // Similar story as the gun looting var.
+
+	// Angles
+	// Remember that in BYOND, NORTH equals 0 absolute degrees, and not 90.
+	var/traverse = 180 // Determines how wide the turret can turn to shoot things, in degrees. The 'front' of the turret is determined by it's dir variable.
+	var/leftmost_traverse = null // How far left or right the turret can turn. Set automatically using the above variable and the inital dir value.
+	var/rightmost_traverse = null
+	var/current_bearing = 0 // Current absolute angle the turret has, used to calculate if it needs to turn to try to shoot the target.
+	var/target_bearing = 0 // The desired bearing. If the current bearing is too far from this, the turret will turn towards it until within tolerence.
+	var/bearing_tolerence = 5 // Degrees that the turret must be within to be able to shoot at the target.
+	var/turning_rate = 90 // Degrees per second.
+
+	// Detection.
+	var/datum/proximity_trigger/square/proximity = null
+	var/vision_range = 7	// How many tiles away the turret can see. Values higher than 7 will let the turret shoot offscreen, which might be unsporting.
+							// Higher values may also have a performance cost.
+	
+	// Logic
+	var/datum/state_machine/turret/state_machine = null
+	var/atom/target = null
```

Comment:
Would probably be worth setting this as a weakref instead of a hard reference - one of the big contributors to qdel failures is this kind of var.

Reply:
````
var/weakref/target = weakref(some_atom)
var/atom/some_atom = target?.resolve()
````